feat: add configmaps & secrets to kubernetes

cdrage opened the pull request:
feat: add configmaps & secrets to kubernetes

### What does this PR do?

* Adds ConfigMaps and Secrets to Kubernetes
* Ability to delete / view / edit both configmap and secrets

### Screenshot / video of UI

<!-- If this PR is changing UI, please include
screenshots or screencasts showing the difference -->
![Screenshot 2024-06-17 at 4 24 23 PM](https://github.com/containers/podman-desktop/assets/6422176/b348e49c-142e-429e-ad25-08a6fe389cf6)

https://github.com/containers/podman-desktop/assets/6422176/41108199-7e2b-4cc2-81f6-2b267b1c887a




### What issues does this PR fix or reference?

<!-- Include any related issues from Podman Desktop
repository (or from another issue tracker). -->

Closes https://github.com/containers/podman-desktop/issues/7342
Closes https://github.com/containers/podman-desktop/issues/7190 (its the last part of it)

### How to test this PR?

<!-- Please explain steps to verify the functionality,
do not forget to provide unit/component tests -->

- [X] Tests are covering the bug fix or the new feature

1. Create a ConfigMap and Secrets on your k8s cluster.
2. Delete / Edit / View them

Signed-off-by: Charlie Drage <charlie@charliedrage.com>

cdrage commented:
FYI the icon is the "rocket" temporary until #7680 has a PR.

Apologizes for the large PR! Tried splitting it up a bit, but a lot of the tests are intertwined.

Most of this code is very similar to #7640 however.

deboer-tim reviewed (commented):
Code worked fine for me in testing and added a few comments, but this PR is so massive that I haven't gone through it all. Parts of this really should be broken out so that we can review and merge independently.

cdrage commented:
@deboer-tim I've updated the PR with a new commit with the suggested changes:

- Added isConfigMap / made the if statements more intuitive
- Use green for the "running" / working configmap and secret.
- Updated the incorrect filtered areas


Still unsure what to do with Label? I guess another PR / issue, since removing the capitalization will affect other parts of PD.

deboer-tim commented on packages/renderer/src/lib/configmaps-secrets/configmap-secret-utils.ts:
```diff
@@ -0,0 +1,56 @@
+
+  // If it is a configMap, then it will **not** have a type property
+  isConfigMap(storage: V1ConfigMap | V1Secret): storage is V1ConfigMap {
+    return !('type' in storage);
```
It would probably be better to have a 'positive' check for something unique about config map vs 'it's not a secret' 😆. Is there nothing that's only in configmap?

cdrage replied:
actually yes! I'll update it to also check for 'binaryData' too.

deboer-tim commented on packages/renderer/src/lib/configmaps-secrets/ConfigMapSecretList.svelte:
```diff
@@ -0,0 +1,189 @@
+          if (isSecret) {
+            await window.kubernetesDeleteSecret(configmapSecret.name);
+          } else {
+            await window.kubernetesDeleteConfigMap((configmapSecret as ConfigMapSecretUI).name);
```
I think you can remove the `as ConfigMapSecretUI` now?

deboer-tim commented on packages/renderer/src/lib/configmaps-secrets/configmap-secret-utils.ts:
```diff
@@ -0,0 +1,56 @@
+/**********************************************************************
+ * Copyright (C) 2023 Red Hat, Inc.
```
Should be 2024

feloy commented:
![configmaps-list](https://github.com/containers/podman-desktop/assets/9973512/af7fdf66-1325-443f-a6e6-19318af17246)

The values in the Type column for secrets are the ones displayed in `kubectl get secret`, but I would prefer to have the ones used when using `kubectl create secret <type>` (docker-registry, generic or tls). WDYT?

For specific secrets (docker, tls), I wonder if we need to display the keys, as they are standard (and not really useful to the user).

`kubectl get cm / secret` display the number of data, without the names of the keys. I wonder if it is relevant to display the key names in the list, as it can easily be unreadable

feloy commented:
![configmap-details](https://github.com/containers/podman-desktop/assets/9973512/319cdc2b-eccd-414c-914d-2e7395c4e86e)

I would prefer to have the display of the data moved up one level:

```
Details

Type      Opaque
Immutable  No


Data

key1       value1
key2       value2
key3       value3
key4       value4
key5       value5
```

cdrage commented:
Thank you for the suggestions @feloy I agree. Having keys on there is not ideal. Other GUI's do it, but because of our spacin and simplicity it's nice to say the length / how many rather than every name of the key..

I've updated the UI with the "Summary" changes as well as the list changes. I agree it looks a lot better.

![Screenshot 2024-06-19 at 2 10 40 PM](https://github.com/containers/podman-desktop/assets/6422176/952e881f-feb0-4400-bb47-d4fb2fc85b9d)
![Screenshot 2024-06-19 at 2 10 54 PM](https://github.com/containers/podman-desktop/assets/6422176/50e11d72-3e0b-4b68-98f9-fec6ae5ba9e5)

feloy commented on packages/renderer/src/lib/kube/details/KubeSecretArtifact.svelte:
```diff
@@ -21,12 +23,13 @@ export let artifact: V1Secret | undefined;
+      <Subtitle>Data</Subtitle>
+    </tr>
+    <tr>
+      {#each Object.entries(artifact.data) as [key, value]}
+        <Cell>{key}</Cell>
+        <Cell>{value}</Cell>
+      {/each}
```
All the keys appear in the same line

cdrage commented:
![Screenshot 2024-06-19 at 11 50 34 AM](https://github.com/containers/podman-desktop/assets/6422176/d45d7280-72bf-401c-9b3e-f077f476f34d)
@feloy updated

cdrage commented:
![Screenshot 2024-06-19 at 4 36 14 PM](https://github.com/containers/podman-desktop/assets/6422176/bf1ccdb5-34c8-4715-bc1b-9b51e9e25320)
Updated so now they are on newlines @feloy

deboer-tim commented on packages/renderer/src/lib/ui/Label.svelte:
```diff
@@ -3,12 +3,13 @@ import { Tooltip } from '@podman-desktop/ui-svelte';
 
 export let name = '';
 export let tip = '';
+export let capitalize = true;
```
This was delivered via #7788, can remove it here.

cdrage commented:
Rebased against the label changes and ready for review @feloy @deboer-tim 

I did not make any changes other than the Label conflict.

deboer-tim commented:
> Rebased against the label changes and ready for review @feloy @deboer-tim
> 
> I did not make any changes other than the Label conflict.

Build failing. TBH this is still a massive PR and I would still feel more comfortable if you threw the main/preload/store in another PR first, as I think those could be merged easy and would cut down the size/focus of this a lot.

cdrage commented:
> > Rebased against the label changes and ready for review @feloy @deboer-tim
> > I did not make any changes other than the Label conflict.
> 
> Build failing. TBH this is still a massive PR and I would still feel more comfortable if you threw the main/preload/store in another PR first, as I think those could be merged easy and would cut down the size/focus of this a lot.

No problem! Going to close this and split it into 1 for API / backend, other for UI.
